PS-2968 Add DateTimes

joneshf-cn opened the pull request:
## What it does

The change here give us support for DateTimes. We're blocked on finishing this by having `UserCleared` exist. #14 is one way to get unblocked.

## Review asks

Did I miss anything? This seemed to be quite a bit easier to implement than currency or dropdowns.

joneshf-cn committed:
PS-2968 Inline `DropdownQuery _`

`DropdownQuery _` doesn't seem to pass the Fairbairn threshold.
It's probably better if we don't start a pattern of naming
things that are only used once.

joneshf-cn committed:
PS-2968 Use infix operators instead of prefix names

We're about to add more children.
Using the prefix names means having to remember to update them
for each child we add.
Rather than add excess noise to our changes,
we opt for the infix operators which should give us mostly signal.

joneshf-cn committed:
PS-2968 Add support for datetimes

This is most of the work needed to have datetime support in our forms.
There is one point we haven't finished yet:
we need to handle the datetime being unset.
The change necessary for that is available here:
https://github.com/citizennet/purescript-lynx/pull/14.

Once that PR is merged (or we somehow end up with `UserCleared`),
we can finish this up and merge this stuff in.

davezuch commented on src/Data/Form.purs:
```diff
@@ -184,6 +193,19 @@ eval get page = do
+      placeholder <- evalExpr get input.placeholder
+      required <- evalExpr get input.required
+      value <- traverse (evalExpr get) input.value
+      pure
```
An example of the problem with my validation PR, merging that in after this, we'd easily miss that this isn't being validated

davezuch commented on src/Page/Form.purs:
```diff
@@ -227,3 +247,21 @@ eval = case _ of
+    DateTimePicker.SelectionChanged (Just val) ->
+      eval (UpdateKey key (datetime_ val) a)
+    DateTimePicker.SelectionChanged Nothing -> do
+      let _ = todo $ SProxy :: SProxy "Handle the removal of a datetime. Will require `UserCleared`."
```
This is neat, I like it

davezuch reviewed (approved):
Looks good! None of my comments are important

joneshf-cn committed:
Merge remote-tracking branch 'origin/master' into HEAD

joneshf-cn committed:
Handle `UserCleared`

In some cases, `DateTimePicker` can send us a `Nothing`.
It seems like it happens if it cannot parse the datetime.
We handle that case by clearing the value.
